Enable per-root module imports

aduth opened the pull request:
Closes: #380

This pull request seeks to enable importing by assuming a root path of each entry in the Webpack configuration. This turned out to be hilariously difficult to achieve, eventually resulting in the implementation of a custom Webpack plugin, currently inlined in the Webpack configuration.

__Example:__

_Before_

```js
import Dashicon from '../../components/dashicon';
```

_After:_

```js
import Dashicon from 'components/dashicon';
```

__Open questions:__

- We could simplify this by merely defining `resolve.modules = [ 'editor', 'i18n', 'element', 'blocks', 'node_modules' ]`, with the downside being that it becomes possible to `import Editable from 'components/editable';` (a `blocks` component) from the `editor` path.
- Should we move the Webpack plugin elsewhere? A separate repository? A standalone file?

__Testing instructions:__

Verify that `npm run build` results in no module resolution errors.

aduth committed:
Add EntryResolvePlugin Webpack plugin

aduth committed:
Convert relative import path traversal to per-module-root

nylen commented on webpack.config.js:
```diff
@@ -2,10 +2,47 @@
+					resolver.plugin( 'module', ( request, callback ) => {
+						// Find entry root which contains the requesting path
+						const resolvePath = entryRoots.find( ( entryRoot ) => {
+							return request.path.startsWith( entryRoot );
```
If I'm understanding this correctly, it would break if you had two directories named like `prefix` and `prefix-and-more-stuff` - there should be a test for a path separator somewhere.

nylen commented:
I'm surprised that something like this doesn't exist yet.  It seems like it would be pretty useful as a standalone plugin.  As a step in that direction, I think it would be good to pull the class out into a separate file in this PR.

This also seems complicated enough that it would benefit from some test cases (for example, for the case I described in https://github.com/WordPress/gutenberg/pull/386#discussion_r110722071).  Have you given much thought to how difficult that would be to achieve (probably easier if it's packaged as a standalone library)?  I found this article on the subject:  https://blog.iansinnott.com/testing-webpack-plugins/

aduth commented:
@nylen Thanks for the feedback. I think I'm going to pull this out as a separate module and give a go at incorporating at least some basic testing.

aduth committed:
Reference ResolveEntryModulesPlugin as newly externalized module

aduth commented:
Extracted the Webpack plugin to an external module and updated the branch accordingly.

https://github.com/aduth/resolve-entry-modules-webpack-plugin
https://www.npmjs.com/package/resolve-entry-modules-webpack-plugin
